dynamo/preprocessing/gene_selection.py

```diff
@@ -0,0 +1,1041 @@
+
+    if mode == "dispersion":
+        if adata.uns[key] is None:
+            main_warning("dispersion mode is deprecated. This mode will be removed in the future.")
+            estimate_dispersion(adata, layers=[layer])
+
+        if adata.uns[key] is None:
+            raise KeyError(
+                "Error: for adata.uns.key=%s, no dispersion model found. Please call estimate_dispersion() before calling this function"
+                % key
+            )
+
+        top_df = pd.DataFrame(
+            {
+                "gene_id": adata.uns[key]["disp_table"]["gene_id"],
+                "mean_expression": adata.uns[key]["disp_table"]["mu"],
+                "dispersion_fit": adata.uns[key]["disp_func"](adata.uns[key]["disp_table"]["mu"]),
+                "dispersion_empirical": adata.uns[key]["disp_table"]["disp"],
+            }
+        )
+        top_df = top_df.set_index("gene_id")
+
+    elif mode == "gini":
+        top_df = adata.var[layer + "_gini"]
+
+    return top_df
+
+
+def select_genes_monocle(
+    adata: AnnData,
+    layer: str = "X",
+    keep_filtered: bool = True,
+    n_top_genes: int = 2000,
+    sort_by: Literal["gini", "cv_dispersion", "fano_dispersion"] = "cv_dispersion",
+    exprs_frac_for_gene_exclusion: float = 1,
+    genes_to_exclude: Union[List[str], None] = None,
+    SVRs_kwargs: dict = {},
+):
+    """Select genes based on monocle recipe.
+
+    This version is here for modularization of preprocessing, so that users may try combinations of different
+    preprocessing procedures in Preprocessor.
+
+    Args:
+        adata: an AnnData object.
+        layer: The data from a particular layer (include X) used for feature selection. Defaults to "X".
+        keep_filtered: Whether to keep genes that don't pass the filtering in the adata object. Defaults to True.
+        n_top_genes: the number of top genes based on scoring method (specified by sort_by) will be selected as feature
+            genes. Defaults to 2000.
+        sort_by: the sorting methods, either SVR, dispersion or Gini index, to be used to select genes. Defaults to
+            "SVR". TODO: Should be fixed!
+        exprs_frac_for_gene_exclusion: threshold of fractions for high fraction genes. Defaults to 1.
+        genes_to_exclude: genes that are excluded from evaluation. Defaults to None.
+        SVRs_kwargs: kwargs for `SVRs`. Defaults to {}.
+
+    Returns:
+        The adata object with genes updated if `only_bools` is false. Otherwise, the bool array representing selected
+        genes.
+    """
+
+    # The following size factor calculation is a prerequisite for monocle recipe preprocess in preprocessor.
+    adata = calc_sz_factor(
+        adata,
+        total_layers=adata.uns["pp"]["experiment_total_layers"],
+        scale_to=None,
+        splicing_total_layers=False,
+        X_total_layers=False,
+        layers=adata.uns["pp"]["experiment_layers"],
+        genes_use_for_norm=None,
+    )
```

Comment:
We have fixed the code based on your suggestion, but noticed that the result has changed since we now perform normalization before gene selection. It seems that further discussion is necessary...

Reply:
Done.

# The following size factor calculation is a prerequisite for monocle recipe preprocess in preprocessor.
        self._calc_size_factor(adata)
        self._normalize_by_cells(adata)
        self._select_genes(adata)

        # append/delete/force selected gene list required by users.
        self._append_gene_list(adata)
        self._exclude_gene_list(adata)
        self._force_gene_list(adata)

        self._log1p(adata)
        self._pca(adata)
        self._cell_cycle_score(adata)